zerver/lib/push_notifications.py

```diff
@@ -657,6 +666,18 @@ def get_remove_payload_gcm(
     gcm_options = {'priority': 'normal'}
     return gcm_payload, gcm_options
 
+def get_remove_payload_apns(user_profile: UserProfile, message_ids: List[int]) -> Dict[str, Any]:
+    zulip_data = get_base_payload(user_profile)
+    zulip_data.update({
+        'event': 'remove',
+        'zulip_message_ids': ','.join(str(id) for id in message_ids),
```

Comment:
`message_ids` seems like it'd be a more consistent name for this.  Also, what's the thinking around marshalling with a string join, rather than keeping it a list of integers (or if we have to, JSON dumping that?)

Reply:
For better or worse, `zulip_message_ids` is the name we already use in the Android case. Here's the comment about why:
```
        'zulip_message_id': message.id,  # message_id is reserved for CCS
```
(And then in the `remove` case we matched that with `zulip_message_id` and later `zulip_message_ids`.)

I don't know what "CCS" is. The comment apparently dates to 662edc255, from 2013. I suspect whatever it's about is no longer relevant; [in the current docs](https://firebase.google.com/docs/cloud-messaging/concept-options#data_messages) the reserved keys 'include "from", "notification," "message_type", or any word starting with "google" or "gcm."'

It is awkward that, already in the non-`remove` case, we say `message_ids` over APNs but `zulip_message_ids` over FCM.

Reply:
> Also, what's the thinking around marshalling with a string join, rather than keeping it a list of integers (or if we have to, JSON dumping that?)

This is also what we already do in the Android / FCM case. With FCM, the data we provide for the actual message as transmitted consists of [key-value pairs](https://firebase.google.com/docs/cloud-messaging/concept-options), and they aren't 100% unambiguous about this but that sounds to me like the values are strings. So even if some layer in between would serialize automatically, I think it's best to choose the serialization explicitly to avoid confusion.

With APNs, the payload is ["a JSON dictionary"](https://developer.apple.com/documentation/usernotifications/setting_up_a_remote_notification_server/generating_a_remote_notification). The same doc goes on to say "Your custom keys must have values with primitive types, such as dictionary, array, string, number, or Boolean", so oh, OK, only "primitive types"... but wait, what's "dictionary" and "array" doing there, doesn't that mean the opposite? A link leads to [this doc](https://developer.apple.com/documentation/usernotifications/unnotificationcontent/1649869-userinfo) for how it shows up in Swift at the client, and that says the data must be "property-list types—that is, they must be types that can be serialized into the property-list format", which [turns out to](https://developer.apple.com/library/archive/documentation/Cocoa/Conceptual/PropertyLists/AboutPropertyLists/AboutPropertyLists.html#//apple_ref/doc/uid/10000048i-CH3-54303) roughly coincide with what you can put in JSON. I think the bottom line is *probably* that it's arbitrary JSON, and there's some confusion in the writing of these docs.

Given that, I'd be inclined to leave it as a list in the APNs case, and let it get automatically serialized and deserialized as JSON.

Reply:
> I don't know what "CCS" is. The comment apparently dates to [662edc2](https://github.com/zulip/zulip/commit/662edc2558447ba5d5ba8ae9c7f918c7e70dd459), from 2013. I suspect whatever it's about is no longer relevant;

Oh wow, some traces on the web (like [this Haskell library doc](http://hackage.haskell.org/package/push-notify-ccs-0.1.0.1/docs/Network-PushNotify-Ccs.html)) suggest it was an old name for [the XMPP protocol](https://firebase.google.com/docs/cloud-messaging/xmpp-server-ref) for talking to what's now FCM.

And when I was looking at the docs a few minutes ago, that was in fact the one place that I spotted references to `message_id` as a special name!
>  message_id        Required, string
> This parameter uniquely identifies a message in an XMPP connection.

So, that's probably it. Apparently in 2013 the way we sent notification messages to Google (and presumably at the time the preferred, latest way to do it) was over XMPP.